### ISSUE TITLE
Consult POA team on bridge with large number of validators

### ISSUE THREAD

**Issue by jannikluhn** [trustlines-protocol/blockchain#37]
The POA bridge seems to typically be deployed with a maximum of maybe 20 validators, but we potentially will need 123. We should contact the developers and ask if they think this could be a problem and if they have thought about optimizations such as batching.

**Comment by jannikluhn** [comment]
If this really becomes an issue, batching could also be done independently of the bridge validators in a trustless fashion. The general idea is to funnel the tokens through two contracts, one on the main and one on the side chain. The input funnel needs to know the output funnel address (as it has to tell the bridge where to send the tokens), and the output funnel needs to know the participants (as it has to distribute the tokens to them). A naive implementation would

1) collect payout addresses
2) deploy the output contract
3) deploy the input contract
4) wait for all participants to pay the input contract
5) send the tokens to the bridge

The first issue with this is that it requires a fair bit of coordination to agree on the set of payout addresses. The second issue is that it can be sabotaged by a single participant by just not sending their tokens to the input contract.

But `CREATE2` allows us to circumvent this: The input contract is long lived, i.e. independent of the current batch. Participants send their tokens to the input contract without knowing who else is or will be participating (i.e., no coordination is required). Anyone can trigger sending an arbitrary set of inputs to the bridge (i.e., no one can block transfers). 

The output funnel contract address is neither specified by the participants, nor the guy who triggers the transfer. Instead, it is calculated by the contract from the set of participants. The actual contract will be created by a factory contract using the `CREATE2` opcode. This ensures that the output contract address is deterministic (dependent on the contract init code, i.e. on the set of participants) and allows the (main chain) input contract to calculate it. This guarantees that the tokens are sent to a contract that (only) the participants can withdraw from, without requiring them to deploy the contract and know its address in advance.

The only thing that could happen is that the tokens are sent to the output contract, but no one bothers to deploy it. Users often probably can't as they don't have access to side chain native tokens. But, we only need a single honest party to do this and we could even incentivize it by giving him a fee, so this shouldn't be a problem

**Comment by jannikluhn** [comment]
Opened a thread in the POA.network forum asking about potential pitfalls when using the bridge with many validators: https://forum.poa.network/t/bridge-with-many-validators/2376

**Comment by jannikluhn** [comment]
Outcome: There's no fundamental problem, the only one is gas costs. Fortunately, on the side chain those can be distributed over many blocks as validators submit their signatures in separate transactions.

According to [this table](https://github.com/poanetwork/poa-bridge-contracts/blob/46151629adb3dd20e6fb47d4ba91b88c6d388c4f/GAS_CONSUMPTION.md), the validator transcations cost between 64380 and 140744 units of gas each (`executeAffirmation`). Probably all except for the last one should be on the lower end, but lets assume is the maximum for all of them. Of the 100 validators, 50 have to sign. So one transfer would cost a total of 7M gas which is a lot but below typical block gas limits. Moreover, if users are fine with waiting for 10 minutes, this would spread over 120 blocks, so we'd have an average gas usage of 60k gas per block.

Looking at it differently: A Trustlines transaction costs about 100k gas as far as I know. So a bridge is equivalent of about 70 Trustlines transactions.

Another thing we have to consider is total throughput of the bridge. If we've got a block gas limit of 10M, we can put ~70 validator transactions into a single block, so the bridge would have a throughput of ~0.3/s or ~25k per day.

My interpretation of all of this is that it should be fine as bridge transfers are a one time event per user. If it is not though we need to look into batching either by the approach I described above, or by large changes to/a rewrite of the bridge's code (both contracts and validator clients).
